grpc_adapter.go

```diff
@@ -0,0 +1,668 @@
+	for key, vals := range header {
+		if strings.HasPrefix(key, http.TrailerPrefix) {
+			// Must remove trailer prefix before canonicalizing.
+			key = http.CanonicalHeaderKey(key[len(http.TrailerPrefix):])
+			trailer[key] = vals
+		} else if key := http.CanonicalHeaderKey(key); isTrailer[key] {
+			trailer[key] = vals
+		}
+	}
+
+	protobuf := w.config.Codec
+	trailerErr := grpcErrorFromTrailer(w.config.BufferPool, protobuf, trailer)
+	if trailerErr != nil {
+		if w.isStreaming() {
+			trailerErr.meta = trailer
+		}
+		return trailer, trailerErr
+	}
+
+	// Remove all protocol trailer keys.
+	for key := range trailer {
+		if isProtocolHeader(key) {
+			delete(trailer, key)
+		}
+	}
+	return trailer, nil
+}
+
+func (w *connectResponseWriter) finalizeUnary() error {
+	trailer, err := w.getTrailer()
+	if err != nil {
+		return err
+	}
+
+	// Encode trailers as header
+	responseHeader := w.ResponseWriter.Header()
+	for key, vals := range trailer {
+		if isProtocolHeader(key) {
+			continue
+		}
+		key = connectUnaryTrailerPrefix + key
+		responseHeader[key] = vals
+	}
+
+	_ = w.body.Next(5) // Discard envelope head.
+	contentLength := strconv.Itoa(w.body.Len())
+	setHeaderCanonical(responseHeader, headerContentLength, contentLength)
+
+	// Write buffered unary body to stream.
+	if _, err := w.body.WriteTo(w.ResponseWriter); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *connectResponseWriter) finalizeStream() error {
+	trailer, err := w.getTrailer()
+	if err != nil {
+		return err
+	}
+
+	// Encode as connect end message.
+	end := &connectEndStreamMessage{
+		Trailer: trailer,
+	}
+	data, err := json.Marshal(end)
+	if err != nil {
+		return err
+	}
+
+	head := [5]byte{connectFlagEnvelopeEndStream, 0, 0, 0, 0}
+	binary.BigEndian.PutUint32(head[1:5], uint32(len(data)))
+	if _, err := w.ResponseWriter.Write(head[:]); err != nil {
+		return err
+	}
+	if _, err := w.ResponseWriter.Write(data); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *connectResponseWriter) flushWithTrailers() {
+	if w.isStreaming() {
+		if err := w.finalizeStream(); err != nil {
+			responseHeader := w.ResponseWriter.Header()
+			setHeaderCanonical(responseHeader, headerContentType, w.typ)
+			_ = w.config.ErrorWriter.writeConnectStreaming(w.ResponseWriter, err)
+		}
+	} else {
+		if err := w.finalizeUnary(); err != nil {
+			responseHeader := w.ResponseWriter.Header()
+			delHeaderCanonical(responseHeader, headerContentLength)
+			delHeaderCanonical(responseHeader, connectUnaryHeaderCompression)
+			setHeaderCanonical(responseHeader, headerContentType, connectUnaryContentTypeJSON)
+			_ = w.config.ErrorWriter.writeConnectUnary(w.ResponseWriter, err)
+		}
+	}
+	flushResponseWriter(w.ResponseWriter)
+}
+
+func (w *connectResponseWriter) Flush() {
+	if w.isStreaming() {
+		if !w.wroteHeader {
+			w.writeHeader(http.StatusOK)
+		}
+		flushResponseWriter(w.ResponseWriter)
+		return
+	}
+}
+
+// NewGRPCAdapter returns a new http.Handler that transcodes connect and
```

Comment:
Style-wise, exported elements are supposed to appear before unexported elements. This has things in the opposite order. Even outside of our [style guide](https://www.notion.so/bufbuild/Go-Code-Standards-ff40382e6b104c1097a017f9565d54db#c6aeeed92adf40e189d7323ba0ccd11d), I find it personally easier to read to see the most important parts -- the entry to the story -- at the top, and then the supporting details below.

Reply:
Thanks, agree.